server/src/controllers/course.ts

```diff
@@ -106,12 +109,130 @@ export async function addCourse(req: Request, res: Response): Promise<void> {
   }
 }
 
+enum GradingType {
+  PassFail = 'PASSFAIL',
+  Numerical = 'NUMERICAL'
+}
+
+enum Period {
+  I = 'I',
+  II = 'II',
+  III = 'III',
+  IV = 'IV',
+  V = 'V'
+}
+
+enum TeachingMethod {
+  Lecture = 'LECTURE',
+  Exam = 'EXAM'
+}
+
+interface CourseInstanceAddRequest {
+  gradingType: GradingType;
+  startingPeriod: Period;
+  endingPeriod: Period;
+  teachingMethod: TeachingMethod;
+  responsibleTeacher: number;
+  startDate: Date;
+  endDate: Date;
+}
+
+const courseInstanceAddRequestSchema: yup.AnyObjectSchema = yup.object().shape({
+  gradingType: yup
+    .string()
+    .oneOf([GradingType.PassFail, GradingType.Numerical])
+    .required(),
+  startingPeriod: yup
+    .string()
+    .oneOf([Period.I, Period.II, Period.III, Period.IV, Period.V])
+    .required(),
+  endingPeriod: yup
+    .string()
+    .oneOf([Period.I, Period.II, Period.III, Period.IV, Period.V])
+    .required(),
+  teachingMethod: yup
+    .string()
+    .oneOf([TeachingMethod.Lecture, TeachingMethod.Exam])
+    .required(),
+  responsibleTeacher: yup
+    .number()
+    .required(),
+  startDate: yup
+    .date()
+    .required(),
+  endDate: yup
+    .date()
+    .required()
+});
+
+export async function addCourseInstance(req: Request, res: Response): Promise<Response> {
+  try {
+    const courseId: number = Number(req.params.courseId);
+
+    await courseInstanceAddRequestSchema.validate(req.body, { abortEarly: false });
+
+    const request: CourseInstanceAddRequest = req.body;
+
+    const course: Course | null = await models.Course.findOne({
+      where: {
+        id: courseId
+      },
+    });
+
+    if (course == null) {
+      throw new Error(`Course with ID ${courseId} does not exist`);
+    }
+
+    const teacher: User | null = await models.User.findOne({
+      where: {
+        id: request.responsibleTeacher
+      },
+    });
+
+    // TODO: Also check that user has correct role
+    if (teacher == null) {
+      throw new Error(`Teacher with ID ${request.responsibleTeacher} does not exist`);
+    }
+
+    const newInstance: CourseInstance = await models.CourseInstance.create({
+      courseId: courseId,
+      gradingType: request.gradingType,
+      startingPeriod: request.startingPeriod,
+      endingPeriod: request.endingPeriod,
+      teachingMethod: request.teachingMethod,
+      responsibleTeacher: request.responsibleTeacher,
+      startDate: request.startDate,
+      endDate: request.endDate
+    });
+
+    return res.send({
+      success: true,
+      instance: newInstance
```

Comment:
I wonder what kind of data is sent to the frontend here. It gives the impression that the data is just echoed here, which feels slightly redundant.

Reply:
I made this edit, so I'll comment. Yes basically its bit redundant since it returns all the data back to client. Only new data included is the instance `id` in db and `createdAt` and `updatedAt` values. If client does not need these values then we could just return the `success: true` or just include the data client does not have:

```JavaScript
return res.send({
  success: true,
  instance: {
    id:  newInstance.id
  }
})
```

Reply:
I just noticed that this is returning a Sequelize model too, do you think any issues could arise from doing that?

Reply:
Looking at the [Sequelize docs](https://sequelize.org/api/v6/class/src/model.js~model), it would seem that the clas has some properties such as `isNewRecord` and `sequelize`, and overall I just get the idea that it's not intended for serialization as-is. I feel it would be best to only return the id to the client, perhaps also the creation/update date. What are the requirements of the frontend side?

Reply:
I changed it to return just the ID of the new instance as suggested above, in case the client needs more information this can be updated later.